Can't compile bitcoin for SmartOS - recipe for target 'leveldb/util/leveldb_libleveldb_a-env_posix.o' failed

<!-- This issue tracker is only for technical issues related to Bitcoin Core.

General bitcoin questions and/or support requests are best directed to the Bitcoin StackExchange at https://bitcoin.stackexchange.com.

For reporting security issues, please read instructions at https://bitcoincore.org/en/contact/.

If the node is "stuck" during sync or giving "block checksum mismatch" errors, please ensure your hardware is stable by running memtest and observe CPU temperature with a load-test tool such as linpack before creating an issue! -->

<!-- Describe the issue -->
<!--- What behavior did you expect? -->

<!--- What was the actual behavior (provide screenshots if the issue is GUI-related)? -->

<!--- How reliably can you reproduce the issue, what are the steps to do so? -->

<!-- What version of Bitcoin Core are you using, where did you get it (website, self-compiled, etc)? -->

<!-- What type of machine are you observing the error on (OS/CPU and disk type)? -->

<!-- Any extra information that might be useful in the debugging process. -->
<!--- This is normally the contents of a `debug.log` or `config.log` file. Raw text or a link to a pastebin type site are preferred. -->

After fixing #13581 my build still fails with...

```
leveldb/util/env_posix.cc: In member function 'virtual leveldb::Status leveldb::{anonymous}::PosixSequentialFile::Read(std::size_t, leveldb::Slice*, char*)':
leveldb/util/env_posix.cc:105:51: error: 'fread_unlocked' was not declared in this scope
     size_t r = fread_unlocked(scratch, 1, n, file_);
                                                   ^
leveldb/util/env_posix.cc: In member function 'virtual leveldb::Status leveldb::{anonymous}::PosixWritableFile::Append(const leveldb::Slice&)':
leveldb/util/env_posix.cc:234:66: error: 'fwrite_unlocked' was not declared in this scope
     size_t r = fwrite_unlocked(data.data(), 1, data.size(), file_);
                                                                  ^
leveldb/util/env_posix.cc: In member function 'virtual leveldb::Status leveldb::{anonymous}::PosixWritableFile::Flush()':
leveldb/util/env_posix.cc:251:30: error: 'fflush_unlocked' was not declared in this scope
     if (fflush_unlocked(file_) != 0) {
                              ^
leveldb/util/env_posix.cc: In member function 'virtual leveldb::Status leveldb::{anonymous}::PosixWritableFile::Sync()':
leveldb/util/env_posix.cc:290:30: error: 'fflush_unlocked' was not declared in this scope
     if (fflush_unlocked(file_) != 0 ||
                              ^
Makefile:5044: recipe for target 'leveldb/util/leveldb_libleveldb_a-env_posix.o' failed
make[2]: *** [leveldb/util/leveldb_libleveldb_a-env_posix.o] Error 1
make[2]: Leaving directory '/opt/local/src/bitcoin/src'
Makefile:9824: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/opt/local/src/bitcoin/src'
Makefile:757: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1

```

My current failure point, fwiw
```
# ./configure --with-gui=no --disable-tests  --disable-wallet --disable-pic --disable-shared
...
checking for boostlib >= 1.47.0... yes
checking whether the Boost::System library is available... yes
checking for exit in -lboost_system... yes
checking whether the Boost::Filesystem library is available... no
checking whether the Boost::Program_Options library is available... yes
checking for exit in -lboost_program_options... yes
checking whether the Boost::Thread library is available... no
checking whether the Boost::Chrono library is available... yes
checking for exit in -lboost_chrono... yes
checking for mismatched boost c++11 scoped enums... ok
configure: error: No working boost sleep implementation found.

# pkgin search boost
p5-Type-Tiny-XS-0.012  XS boost for some of Type::Tiny's type constraints
boost-python-1.57.0 = Free, peer-reviewed portable C++ source libraries (Boost.Python binary library)
boost-libs-1.57.0 =  Free, peer-reviewed portable C++ source libraries (binary libraries)
boost-jam-1.57.0 =   Free, peer-reviewed portable C++ source libraries (Boost.Jam utility)
boost-headers-1.57.0 = Free, peer-reviewed portable C++ source libraries (build-time headers)
boost-docs-1.57.0 =  Free, peer-reviewed portable C++ source libraries (documentation)
boost-build-1.57.0 = Free, peer-reviewed portable C++ source libraries (Boost.Build framework)
boost-1.57.0 =       Free, peer-reviewed portable C++ source libraries

=: package is installed and up-to-date
<: package is installed but newer version is available
>: installed package has a greater version than available package
```

@MarcoFalke thanks for the suggestion! Output:

```
[bitcoin/depends]# make
Makefile:72: hosts/solaris2.11.mk: No such file or directory
Makefile:74: builders/solaris2.11.mk: No such file or directory
/bin/sh: line 1: Makefile: not found
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: line 1: Makefile: not found
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: line 1: Makefile: not found
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: line 1: Makefile: not found
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: line 1: Makefile: not found
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: line 1: Makefile: not found
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: line 1: Makefile: not found
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: line 1: Makefile: not found
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: line 1: Makefile: not found
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: line 1: Makefile: not found
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: syntax error at line 1: `|' unexpected
/bin/sh: syntax error at line 1: `|' unexpected
make: config.site.in: Command not found
/bin/sh: syntax error at line 1: `|' unexpected
make: config.site.in: Command not found
/bin/sh: syntax error at line 1: `|' unexpected
make: *** No rule to make target 'builders/solaris2.11.mk'.  Stop.
```

--- a/configure.ac
+++ b/configure.ac
@@ -541,9 +541,16 @@ case $host in
      CPPFLAGS="$CPPFLAGS -DMAC_OSX"
      OBJCXXFLAGS="$CXXFLAGS"
      ;;
+   *android*)
+     dnl make sure android stays above linux for hosts like *linux-android*
+     LEVELDB_TARGET_FLAGS="-DOS_ANDROID"
+     ;;
    *linux*)
      TARGET_OS=linux
      LEVELDB_TARGET_FLAGS="-DOS_LINUX"
+     ;;
+   *kfreebsd*)
+     LEVELDB_TARGET_FLAGS="-DOS_KFREEBSD"
      ;;
    *freebsd*)
      LEVELDB_TARGET_FLAGS="-DOS_FREEBSD"
@@ -554,10 +561,17 @@ case $host in
    *netbsd*)
      LEVELDB_TARGET_FLAGS="-DOS_NETBSD"
      ;;
+   *dragonfly*)
+     LEVELDB_TARGET_FLAGS="-DOS_DRAGONFLYBSD"
+     ;;
+   *solaris*)
+     LEVELDB_TARGET_FLAGS="-DOS_SOLARIS"
+     ;;
+   *hpux*)
+     LEVELDB_TARGET_FLAGS="-DOS_HPUX"
+     ;;
    *)
-     OTHER_OS=`echo ${host_os} | awk '{print toupper($0)}'`
-     AC_MSG_WARN([Guessing LevelDB OS as OS_${OTHER_OS}, please check whether this is correct, if not add an entry to configure.ac.])
-     LEVELDB_TARGET_FLAGS="-DOS_${OTHER_OS}"
+     AC_MSG_ERROR(Cannot build leveldb for $host. Please file a bug report.)
      ;;
 esac
 

